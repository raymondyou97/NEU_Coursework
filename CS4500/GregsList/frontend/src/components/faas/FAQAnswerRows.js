import PropTypes from 'prop-types';
import React, { useRef, useState } from 'react';
import { Button } from 'react-bootstrap';

const FAQAnswerRows = ({ faas, onDelete, onUpdate }) => {
  const [editId, setEditId] = useState(undefined);
  const answerInputRef = useRef(undefined);

  return faas.map(faqAnswer => {
    const editing = editId === faqAnswer.id;

    return (
      <tr key={faqAnswer.id}>
        <td>{faqAnswer.frequentlyAskedQuestion.title}</td>
        <td>{faqAnswer.frequentlyAskedQuestion.question}</td>
        <td>
          {editing ? (
            <input ref={answerInputRef} defaultValue={faqAnswer.answer} />
          ) : (
            faqAnswer.answer
          )}
        </td>
        <td>
          <Button
            onClick={
              editing
                ? // Save the edits to this row
                  () => {
                    onUpdate(faqAnswer.id, {
                      answer: answerInputRef.current.value,
                    });
                    setEditId(undefined);
                  }
                : // Start editing this row
                  () => setEditId(faqAnswer.id)
            }
            size="sm"
            style={{ height: 'auto' }}
          >
            {editing ? 'Save' : 'Edit'}
          </Button>
          <Button
            onClick={() => onDelete(faqAnswer.id)}
            variant="danger"
            size="sm"
            style={{ height: 'auto' }}
          >
            X
          </Button>
        </td>
      </tr>
    );
  });
};

FAQAnswerRows.propTypes = {
  faas: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      answer: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
};

export default FAQAnswerRows;
