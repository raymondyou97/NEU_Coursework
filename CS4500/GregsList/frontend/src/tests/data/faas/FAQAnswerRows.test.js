import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FAQAnswerRows from '../../../components/faas/FAQAnswerRows';
import { findAllFAQAnswers } from './MockFaaService';

test('FAQ Answer Rows', () => {
  const faas = findAllFAQAnswers();

  const component = shallow(<FAQAnswerRows faas={faas} />);
  expect(component.find('tr').length).toEqual(faas.length);

  const firstRow = component.find('tr').first();
  expect(
    firstRow
      .find('td')
      .at(0)
      .text()
  ).toEqual(faas[0].question);
  expect(
    firstRow
      .find('td')
      .at(1)
      .text()
  ).toEqual(faas[0].answer);
});

test('Snapshot renders', () => {
  const faas = findAllFAQAnswers();

  const component = renderer.create(<FAQAnswerRows faas={faas} />);

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});
