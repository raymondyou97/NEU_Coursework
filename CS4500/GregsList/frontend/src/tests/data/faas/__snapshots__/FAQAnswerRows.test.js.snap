// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot renders 1`] = `
Array [
  <tr>
    <td>
      how many employees do you have?
    </td>
    <td>
      we have 300 people.
    </td>
  </tr>,
  <tr>
    <td>
      how many employees do you have?
    </td>
    <td>
      10 people and growing!
    </td>
  </tr>,
  <tr>
    <td>
      how long have you been operating?
    </td>
    <td>
      we've been in business 15 years
    </td>
  </tr>,
  <tr>
    <td>
      how long have you been operating?
    </td>
    <td>
      Just got started last year
    </td>
  </tr>,
  <tr>
    <td>
      where are you based out of?
    </td>
    <td>
      On Mass Ave!
    </td>
  </tr>,
]
`;
